### ISSUE TITLE
dfu-util: No DFU capable USB device available

### ISSUE THREAD

**Issue by rtek1000** [rogerclarkmelbourne/STM32duino-bootloader#68]
Hi,

Using Windows 10 x64 is fine, but using Xubuntu 18.10 x64, I have this error:
(using generic_boot20_pc13_fastboot.bin)

> /home/userz/Arduino/hardware/Arduino_STM32/tools/linux/maple_upload: line 29: /home/userz/Arduino/hardware/Arduino_STM32/tools/linux/upload-reset:  nonexistent file or directory
> dfu-util: Invalid DFU suffix signature
> dfu-util: A valid DFU suffix will be required in a future dfu-util release!!!
> dfu-util 0.8
> 
> Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
> Copyright 2010-2014 Tormod Volden and Stefan Schmidt
> dfu-util: Invalid DFU suffix signature
> This program is Free Software and has ABSOLUTELY NO WARRANTY
> dfu-util: A valid DFU suffix will be required in a future dfu-util release!!!
> Please report bugs to dfu-util@lists.gnumonks.org
> 
> dfu-util: No DFU capable USB device available
> Waiting for /dev/ttyACM0 serial...Done
 
I tried to follow what this link says:
https://github.com/rogerclarkmelbourne/Arduino_STM32/wiki/Installation

> 
> Linux
> 
>     Copy the Arduino_STM32 folder to the hardware folder in your Arduino sketches folder. If the hardware folder does not exist, please create one.
>     Run the udev rules installation script in tools/linux/install.sh.
>     Note. If you are uploading via USB to Serial or STLink etc, you may need to set the relevant permissions for your specific upload device in order to be able to use it from within the Arduino IDE. You may also need to change the udev rules for the device in question.

Note: this part of the rules I do not quite know how it works.

I also found:

[Fix for possible break of DFU functionality #567]
https://github.com/rogerclarkmelbourne/Arduino_STM32/pull/567
(shodan8192 wants to merge 1 commit into rogerclarkmelbourne:master from shodan8192:master)

What can I do to solve this?

Thank you.

**Comment by rtek1000** [comment]
Hi,

I found the explanation: "perform manual reset to upload"
https://www.stm32duino.com/viewtopic.php?f=28&t=3158

About break of DFU, found these posts:
http://www.stm32duino.com/viewtopic.php?f=53&t=4125

So it seems to be failing the reset part, which in Windows goes well, but Linux does not work.

**Comment by rtek1000** [comment]
Hi, using a logic analyzer and an FTDI FT232RL serial port, I was able to get this data, and apparently it should be the reason it works in Windows, but it does not work on Linux.

![wrong time](https://user-images.githubusercontent.com/8660811/56304256-1de07400-6114-11e9-82d6-558e8ddfabff.png)

**Comment by rtek1000** [comment]
Note: I could not make it work with the original upload-reset file, so I recompiled the source code (gcc -o upload-reset upload-reset.c)

For original upload-reset shows error:
> /home/userz/Arduino/hardware/Arduino_STM32/tools/linux/maple_upload: line 29: /home/userz/Arduino/hardware/Arduino_STM32/tools/linux/upload-reset: No such file or directory

**Comment by rtek1000** [comment]
(See that compile time for Linux is really a big advantage, 3 times!)

**Comment by rtek1000** [comment]
Hi,

I did another test, now with python, and it happened as expected, the same behavior does not occur in Windows and Linux. In Windows the display image becomes a black screen. On Linux it just freezes, but there's no black screen.
(Times seem to be slightly different between Python 2 and Python 3, but in both the screen freezes.)

I will try to run this data stream on a non-debian-derived distribution.

Python script:
```
import time
import serial

# For Linux:
# ser = serial.Serial('/dev/ttyUSB0', 9600, serial.EIGHTBITS, serial.PARITY_NONE, serial.STOPBITS_ONE)

# For Windows:
ser = serial.Serial('COM6', 9600, serial.EIGHTBITS, serial.PARITY_NONE, serial.STOPBITS_ONE)
ser.setRTS(False)
ser.setDTR(False)
ser.setDTR(True)

time.sleep(0.05)

ser.setDTR(False)
ser.setRTS(True)
ser.setDTR(True)

time.sleep(0.05)

ser.setDTR(False)

time.sleep(0.05)

ser.write(b'1EAF')
```

**Comment by rtek1000** [comment]
Hi,

Good news (or bad news?)

I ran the python script on the Raspberry Pi 1 card (S.O. 2018-11-13-raspbian-stretch-lite), and the result in the logic was similar to what was obtained in Windows. The display attached to the STM32 has become a black screen, just like in Windows.

With this I believe that an Operating System problem is identified.

**Comment by rtek1000** [comment]
Bad news (or good news?)

I just installed Debian 9, and it did not work, the screen got the image frozen, it did not turn black screen.

![debian 9 - fail](https://user-images.githubusercontent.com/8660811/56339263-d4c80880-6183-11e9-836f-d5d4c97f0493.png)

**Comment by rtek1000** [comment]
Hi,

Well, after several attempts to find out why Linux does not have the same behavior as Windows. I ended up searching for the Maple source code. And there it was: only the DTR is switched before sending the bytes "1AEF" @ 115200 bauds.

I re-tested it, on Windows, it worked. On the Linux (Debian 9) did not work.
In Raspberry Pi it only works once, if STM32 is connected from the beginning of the Raspbian boot (Debian derived).

I thought, why not test on Android? Yes, on Android worked.

It seems that Linux (or the serial port driver) does not perform an internal reset, perhaps on the USB controller.

New Python script:
> import time
> import serial
> ser = serial.Serial('COM7', 115200, serial.EIGHTBITS, serial.PARITY_NONE, serial.STOPBITS_ONE)
> ser.setDTR(True)
> time.sleep(0.1)
> ser.setDTR(False)
> ser.write(b'1EAF')

![Android](https://user-images.githubusercontent.com/8660811/56376547-86942300-61de-11e9-859b-03e2dec6f2a9.png)

For Android (9) I used B4A, code attached.
[felUsbSerial_DFU.zip](https://github.com/rogerclarkmelbourne/STM32duino-bootloader/files/3095295/felUsbSerial_DFU.zip)
